Pre-processing enhancements

AlexandruPascu opened the pull request:
# Description

- Added the option for the user to say where the header data starts in the file, and the cycler the experiment was conducted with.
- Sanity checks are implemented as a tool for better understanding the pre-processing and the data available for further segmentation or processing techniques, besides the logger.
- Changed the handling of the cycler detection to be more robust and be able to pick up the cycler, but affected the time it takes, also improved data header detection.
- Added a new feature to detect the absolute time of the experiment from the experimental data or the metadata of the file which is added as a new column.

Fixes #22  (issue)
Fixes #14  (issue)
Fixes #11  (issue)
Fixes #9  (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/paramm-team/pybamm-param/blob/main/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [ ] No style issues: `$ python -m nox -s lint`
- [ ] All tests pass: `$ python -m nox -s unit examples integration`
- [ ] Coverage is complete `$ python -m nox -s coverage`
- [ ] The documentation builds: `$ python -m nox -s sphinx`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

AlexandruPascu committed:
Updated Parser methods

Implemented sanity checks while refining core functionalities and sorting bugs for various cyclers

AlexandruPascu committed:
Merge branch 'api-updates' of https://github.com/paramm-team/data_processing into api-updates

AlexandruPascu committed:
Merge branch 'dev' into api-updates

AlexandruPascu committed:
Cycler Detection & Absolute Time

Adding the functionality to detect the cycler based on the keywords found and the option for the user to tell it or the line from where data starts and metadata finishes.

Implemented Absolute time detection taking it from the available recorded testing data or the metadata of the experiment/file.

Adjusted the logger for the new changes

brosaplanella committed:
Merge branch 'main' into api-updates

brosaplanella commented on pbdp/pbdp_parser.py:
Does the name need to be `pbdp_parser.py` or can it just be `parser.py`?

AlexandruPascu replied:
Alright, that makes sense. That's smth I don't know actually since I fetched the changes from his work into my branch 2 weeks ago and it was like that

brosaplanella replied:
Then let's see what Pip says about that.

PipGrylls replied:
I had changed that name as parser was colliding/could collide with other items in the namespace. You can go a few ways with it:
1. Keep it as pbdp_parser which is nice and unique so wont hit other ‘parser’ things
2. Keep it as parser but stop it being importable from pbdp i.e. ‘import pbdp; my_parser = pbdp.parser()’ is okay but ‘from pbdp import parser’ would fail and it also would not be imported doing ‘from pbdp import *’ but then again people shouldn’t do that anyway.
3. Give it another name that isn’t parser.

PipGrylls replied:
https://docs.python.org/3.9/library/parser.html

it ‘was/is’ in the python standard library (it is deprecated) but still poses a bit of an issue.


parser is also a common name in the argparse pattern

https://docs.python.org/3/library/argparse.html

brosaplanella replied:
Happy to call it pbdp parser, thanks for the clarification!

brosaplanella reviewed (approved):
This looks good to me (see a minor comment about file names). However, this also includes quite a few of @PipGrylls changes. I will let Pip handle this as part of the release so he can merge things in the right way. Please ping me if I need to review again later on.

PipGrylls reviewed (dismissed):
All looks good from what I can see. 

I need to put a local commit hook to clear metadata from the notebooks as they are not rendering diffs correctly.

PipGrylls committed:
Merge branch 'main' into api-updates

PipGrylls committed:
linting

PipGrylls commented:
Working on it now but most of the tests are failing, if the parser is updated, especially the default cycler e.t.c, the tests need to be updated in lockstep.

PipGrylls committed:
working on tests

PipGrylls commented:
Hi @AlexandruPascu Could you look over the tests, and do a bit of digging into the function defaults you’ve given. 
Some of the positional/default arguments are now None or Empty strings that get checked and crash out later.
The command:

`nox -s unit`

Should pass without issue and until it does we cant merge these branches. I’ve pushed my initial work on it happy to chat on teams/zoom e.t.c...

AlexandruPascu commented:
Hello @PipGrylls! Thanks for having a look over everything! Can you please drop me a message over Teams with when you'd be available for a call tomorrow? I think it would be easier to understand

AlexandruPascu committed:
Sorted unit test error

Did small changes throughout the code to be more consistent with types and checks

AlexandruPascu committed:
Fixed code style

AlexandruPascu committed:
Fixed python 3.8

AlexandruPascu commented:
@PipGrylls could you please have a look at the Ubuntu tests? I think they are trying to run the example while the others not? Please let me know if there is smth I can do about that and have a look again at the code since now it's passing the tests :)

brosaplanella committed:
comment out examples test in workflow
